### PR TITLE
Port to ZYNQ7000 platforms (ARM Cortex A9)

### PR DESCRIPTION
**Pull request by smee204**
This PR provides support for the Xilinx ZYNQ7000 FPGA family. The processor in the ZYNQ FPGA is a dual core ARM Cortex A9 processor. This is the initial commit for issue #3490
This port only uses a single ARM core.

A board configuration has been created for the Zedboard development board.

The following drivers are included:
- UART
- Timers
- Interrupt controller
- GPIO

### CODE REVIEW

**Comment by kaspar030** [comment]
Nice!

Just took a quick look, but the hwtimer implementation is based on timer/periph, right? would it make sense to use hwtimer_compat?

**Comment by smee204** [comment]
Yes it is based on timer/periph.
I wasn't aware that hwtimer_compat existed. I take it that just replaces my manually created hwtimer_arch?

**Comment by OlegHahm** [comment]
@kaspar030, @haukepetersen, @thomaseichinger, any volunteers for reviewing this PR?

File: boards/zedboard/Makefile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# tell the Makefile.base which module to build
+MODULE = $(BOARD)_base
+
+include $(RIOTBASE)/Makefile.base
+
```
Comment (thomaseichinger) [inline boards/zedboard/Makefile]: Please remove extra newlines.

File: boards/zedboard/include/periph_conf.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#endif
+
+/**
+ * @name UART configuration
+ * @{
+ */
+#define UART_NUMOF          (1U)
+#define UART_0_EN           0
+#define UART_1_EN           1
+#define UART_0_DEV         UART0
```
Comment (thomaseichinger) [inline boards/zedboard/include/periph_conf.h]: Please align.

File: boards/zedboard/scripts/create_boot_file.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+MYDIR="$(dirname "$(which "$0")")"
+
+usage()
+{
+cat << EOF
+usage: $0 options
+
+OPTIONS:
+   -h      Show this message
+   -f	   riot elf file
```
Comment (thomaseichinger) [inline boards/zedboard/scripts/create_boot_file.sh]: Please align.

File: boards/zedboard/scripts/dow.tcl [hunk 4]
```diff
@@ -0,0 +1,12 @@
+connect arm hw
+rst
+
+source cache/ws/zed_hw_platform/ps7_init.tcl
+
+ps7_init
+ps7_post_config
+
+#puts bin/$::env(BOARD)/$::env(APPLICATION).elf
```
Comment (thomaseichinger) [inline boards/zedboard/scripts/dow.tcl]: Please remove not needed lines.

File: boards/zedboard/scripts/prog_jtag.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+export WS=cache/ws
+source /opt/Xilinx/SDK/2014.2/settings64.sh
+
+if [ ! -e $WS/zed_hw_platform/ps7_init.tcl ]; then
+
+  echo "generate hw $PWD"  
+
+  rm -fr $WS
+  xsdk -wait -script $MYDIR/my_SDKproj.xml -workspace $WS
+  #xsdk -wait -eclipseargs -nosplash -application org.eclipse.cdt.managedbuilder.core.headlessbuild -build all -data $WS -vmargs -Dorg.eclipse.cdt.core.console=org.eclipse.cdt.core.systemConsole
```
Comment (thomaseichinger) [inline boards/zedboard/scripts/prog_jtag.sh]: Please remove code not used.

File: cpu/zynq7000/Makefile.include [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export CPU_ARCH = cortex-a9
+
+export USEMODULE += ZYNQ7000_PERIPH
+
+# Explicitly tell the linker to link the vectors.
+#   Without this the interrupt vectors will not be linked correctly!
+#ifeq ($(COMMON_STARTUP),)
```
Comment (thomaseichinger) [inline cpu/zynq7000/Makefile.include]: Remove if not needed.

File: cpu/zynq7000/asm_vectors.S [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	push 	{r0-r3, r12}
+	/* ensure stack is 8-byte aligned */
+	and     r1, sp, #4
+	sub     sp, sp, r1
+	/* store adjustment and lr */
+	push    {r1, lr}
+	/* handle interrupt */
+	bl	IRQInterrupt
+	/* restore lr and unadjust stack */
+	pop		{r1, lr}
```
Comment (thomaseichinger) [inline cpu/zynq7000/asm_vectors.S]: Please align, here and following.

File: cpu/zynq7000/hwtimer_arch.c [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        timer_clear(TIMER_3, timer - PROXY_TIMER_0_CHANNELS - PROXY_TIMER_1_CHANNELS - PROXY_TIMER_2_CHANNELS);
+    }
+#endif
+}
+
+unsigned long hwtimer_arch_now(void) {
+#if TIMER_0_EN == 1
+    return timer_read(TIMER_0);
+#else
+#if TIMER_1_EN == 1
```
Comment (thomaseichinger) [inline cpu/zynq7000/hwtimer_arch.c]: Why don't you use `#elif`?

File: cpu/zynq7000/include/periph_cpu.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+typedef uint32_t gpio_t;
+
+/**
+ * @brief   Available GPIO ports
+ */
+enum {
+    BANK0 = 0,
+    BANK1 = 1,
+    BANK2 = 2,
+	BANK3 = 3,
```
Comment (thomaseichinger) [inline cpu/zynq7000/include/periph_cpu.h]: Align please.

File: cpu/zynq7000/include/zynq7000.h [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	uint32_t event_register_2;
+	uint32_t event_register_3;
+}xttc_ps;
+
+/**
+ * @brief Baseaddress for Zynq7000 peripherals
+ */
+#define TTC0        		((volatile xttc_ps     *)0xF8001000U)
+#define TTC1        		((volatile xttc_ps     *)0xF8002000U)
+#define UART0				((volatile xuart_ps    *)0xE0000000U)
```
Comment (thomaseichinger) [inline cpu/zynq7000/include/zynq7000.h]: This line and following, please align.

File: cpu/zynq7000/thread_arch.c [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+		*stk = ~((uint32_t) STACK_MARKER);
+	}
+
+	/* pc - initial program counter value := thread entry function */
+	stk--;
+	*stk = (uint32_t) task_func;
+
+	/* r3 - contents of r3. required for threads starting from thread_arch_yield */
+	stk--;
+	*stk = (uint32_t) 3;
```
Comment (thomaseichinger) [inline cpu/zynq7000/thread_arch.c]: Could you elaborate on this section concerning r3?

**Comment by thomaseichinger** [comment]
@smee204 Great work. Started reviewing but didn't come quite through yet. One minor remark, could you please check your code for alignment and commented code? Will install qemu to give it a run. Does anybody else have the hardware to test? ( @haukepetersen @OlegHahm @emmanuelsearch @PeterKietzmann )

**Comment by OlegHahm** [comment]
@thomaseichinger, thanks for taking over. No, I don't think I have the hardware.

**Comment by haukepetersen** [comment]
Same here: no hardware available for testing

**Comment by emmanuelsearch** [comment]
Don't have hardware either so far. But might able to buy some. What hardware should we test this on ? The Zedboard? Else, is there another (cheaper) alternative?

**Comment by smee204** [comment]
Unfortunately I am moving jobs so I wont be able to contribute to this port any more as I won't have the time or access to hardware. It is unlikely that anyone else in our company will pick this up this year.

I know there was another company interested in this port, hopefully they can take over.

**Comment by smee204** [comment]
The port was written for a Zedboard however it will run with very little change on any Zynq board. I think the ZYBO is the cheapest.

**Comment by OlegHahm** [comment]
@emmanuelsearch, @haukepetersen, @thomaseichinger, @kaspar030, what's your opinion.

@smee204, sorry for the long silence. This got somehow lost.

**Comment by kYc0o** [comment]
It will be really interesting to get this board! ;)

**Comment by thomaseichinger** [comment]
@OlegHahm Taken the fact @smee204 moved jobs and doesn't have access to the hardware as neither do we I would say we close it for now. We could ask on the mailing list if somebody is interested in support for this board and willing to take over however.

**Comment by kYc0o** [comment]
Feature freeze -> postponed. Can you retake this and rebase it? It would be really nice to have at least one FPGA based board.

**Comment by miri64** [comment]
Postponed due to feature freeze

**Comment by kaspar030** [comment]
Many months without update. I'll set the memo flag and close for now.

**Comment by OlegHahm** [comment]
@kYc0o, this is the board that you have in your office, right?

**Comment by kYc0o** [comment]
Yes a very similar. Peripherals are different but the core is the same.

**Pull request closed without merging**